Graphics engine revamp

AtlantiaKing opened the pull request:
Added IMesh interface and DirectXMesh implementation
Added wrappers around IMaterial, ITexture, IMesh and Shaders in the base engine. These wrappers can be created on the stack as const objects
Removed expensive reload on screen size change
Rewrote the terrain component

AtlantiaKing committed:
Optimized MeshRenderer

MeshRenderer now has a dirty flag which is set if the transform has changed, which stops transform calculations for static meshes in the scene

AtlantiaKing committed:
Removed redundant graphics reloads

Only the swapchain, rendertarget, viewport and spriterenderer need to be reloaded when there's a antialiasing or windowsize rescale.
Removed reloads from the ReloadDirectXEngine method and renamed it to CreateDirectXEngine
Added CreateRenderTargetAndSetViewport method which reloads the swapchain, rendertarget, viewport and spriterenderer

AtlantiaKing committed:
Removed all unnecessary reload functions

Removed the reload functions from all the classes that had one

AtlantiaKing committed:
Removed redundant variables

Removed cached variables from DirectXMaterial and DirectXMeshRenderer since they are no longer needed for reloading

AtlantiaKing committed:
Refactored MeshRenderer

Meshes are now seperated into a IMesh interface in the graphics engine
The IMesh class is wrapped inside a Mesh class in the base engine
Changed LoadMesh function with a SetMesh function inside the MeshRenderer classes

AtlantiaKing committed:
Add PrintDiagnostics method to DirectXEngine

Added method to the DirectXEngine that prints the amount of renderers, materials, meshes and textures currently alive

AtlantiaKing committed:
A mesh now removes itself if unused

A mesh now removes itself from the graphics engine if the mesh is never used

AtlantiaKing committed:
Fixed meshes not being removed when meshrenderer is destroyed

The mesh now removes itself from the graphics engine by calling the Remove method
The MeshRenderer now calls the remove method of the mesh class at the right times

AtlantiaKing committed:
Added templated GraphicsObject class

Moved the graphics object and its use counter from the Mesh class to a templated GraphicsObject class
GetInternal method from the Mesh class is now const

AtlantiaKing committed:
Created texture wrapper

Created Texture class inside the base engine that is a wrapper around a texture (loaded from disk, no custom texture support yet)

AtlantiaKing committed:
Fixed memory leaks in DirectXMesh

AtlantiaKing committed:
Created Material and Shader wrapper

AtlantiaKing committed:
Updated Terrain Component with new wrappers

AtlantiaKing committed:
Fixed LightViewProjection matrix not set on some materials

AtlantiaKing committed:
Updated TerrainComponent to use Texture wrapper

ITexture SetData now takes a const void* instead of a void*
ITexture GetSize now returns unsigned vec2 instead of int vec2
DirectXTexture now creates an RGBA texture when calling LoadTexture(width,height) instead of pure grayscale
Fixed cast error in the window class
Added SetTexture method to the Material class
Added OnInternalChange event to the Texture class
Added Load(width,height) and SetData method to the Texture class
Fixed null exception errors in the texture class
GetSize method in the Texture class now returns unsigned vec2 instead of int vec2
Updated TerrainComponent to use a Texture wrapper to sample its terrain, Added GetHeights and SetHeights methods
Fixed all terrain components having the same material

AtlantiaKing committed:
Removed CPU Access flags from LoadTexture method

AtlantiaKing committed:
Fixed terrain runtime error on exit and terrain position not working

Added UnsetMaterial and UnsetMesh methods to the IMeshRenderer and DirectXMeshRenderer to remove a mesh or material from a renderer without removing it from the graphics engine
Fixed terrain runtime error on exit by unsetting the mesh from the meshrenderer if the mesh is still shared with other terrains
Fixed terrain component world transform not being set to the meshrenderer

AtlantiaKing committed:
Fixed texture and terrain scale issues

Fixed GetData returning too many values due to DirectX padding when creating textures
Changed arguments of the ITexture SetData from a void* with nrBytes to a std::vector of unsigned chars
Removed old terrain code that was commented in the TerrainComponent class
Fixed terrains having 1 meter too much terrain which altered because the texture size was too small

AtlantiaKing committed:
Fixed black border on terrains

AtlantiaKing committed:
Added materials removing their textures

AtlantiaKing committed:
Replaced GetTexture method with LoadHeightmap method in TerrainComponent

AtlantiaKing committed:
Added templated create functions to the Material class

AtlantiaKing committed:
Removed redundant move operators

AtlantiaKing committed:
Fixed compile error when using Material Create method

AtlantiaKing committed:
Fixed SetParent not setting the parent before changing transform

AtlantiaKing committed:
Updated main menu scene to showcase revamped graphics engine

AtlantiaKing committed:
Merge branch 'main' into graphics-engine-revamp

AtlantiaKing committed:
Fixed build problems after fix merge conflicts

Added the DrawLine function that was removed by accident during the merge conflict fix in the DirectXEngine class
Added the Physics class to the CmakeLists of the base engine because it was accidentally removed during the merge conflict fix

LeeVangraefschepe commented on Graphics/DirectX/DirectXMesh.h:
```diff
@@ -0,0 +1,36 @@
+	class DirectXEngine;
+	class DirectXMeshRenderer;
+
+	class DirectXMesh final : public IMesh
```
No rule of 5 applied on non default destructor

LeeVangraefschepe commented on Graphics/DirectX/DirectXMeshRenderer.h:
```diff
@@ -36,28 +37,20 @@ namespace leap::graphics
-		unsigned int m_NrIndices{};
-		ID3D11Buffer* m_pIndexBuffer{};
-
 		bool m_HasCustomMesh{};
```
Unused boolean

LeeVangraefschepe commented on LeapEngine/Components/RenderComponents/TerrainComponent.cpp:
```diff
@@ -1,121 +1,161 @@
-	ApplyTexture();
+leap::TerrainComponent::~TerrainComponent()
+{
+	auto& meshData{ m_Meshes[m_Size] };
```
Searching map with unsigned int while the map is a int

LeeVangraefschepe commented on LeapEngine/Components/RenderComponents/TerrainComponent.cpp:
```diff
@@ -1,121 +1,161 @@
+	--meshData.useCounter;
+	if (meshData.useCounter == 0)
+	{
+		m_Meshes.erase(m_Size);
```
Same here

LeeVangraefschepe commented on LeapEngine/Components/RenderComponents/TerrainComponent.cpp:
```diff
@@ -1,121 +1,161 @@
-	pMeshRenderer->SetMaterial(pMaterial);
-}
\ No newline at end of file
+	m_Meshes[size] = TerrainMesh{ 0, std::move(mesh) };
```
Same here

LeeVangraefschepe commented on LeapEngine/Components/RenderComponents/TerrainComponent.h:
```diff
@@ -2,10 +2,59 @@
+	};
+
+
+	/*class MeshRenderer;
```
Remove unused commented code

LeeVangraefschepe commented on LeapEngine/Graphics/GraphicsObject.h:
```diff
@@ -0,0 +1,55 @@
+namespace leap
+{
+	template<typename T>
+	class GraphicsObject final
```
Wrapper that has the same functionality as a shared pointer? Make this a general class and place it in Utils.

LeeVangraefschepe commented on LeapEngine/Graphics/Shader.cpp:
Empty file?

LeeVangraefschepe reviewed (changes requested):
The ownership of meshes, materials and textures in the engine itself will work. But the code looks very cluttered in DirectX engine (can be changed in later pr).

I also was going to comment on the default cached = true not being a thing in the graphics engine but this is wrapped in leap engine itself.

Look at the comments for more specific feedback.

AtlantiaKing committed:
Applied code review

Added rule of five to the DirectXMesh class
Removed unused variable from DirectXMeshRenderer class
Removed redundant cpp file of header only class
Removed comments from the TerrainComponent class
Changed unorderedmap in the TerrainComponent from int as keys to unsigned int as keys

AtlantiaKing committed:
Added rule of 5 to the GraphicsObject class

AtlantiaKing committed:
Added move operators and setters to material class

LeeVangraefschepe reviewed (approved):
Feedback is applied
